File: src/minima_harness/agent/events.py
"""Agent runtime events — a port of PI's ``pi-agent-core`` event taxonomy.

Emitted by :func:`minima_harness.agent.loop.agent_loop` in a strict order per turn::

    agent_start
      (per turn)
        turn_start
        message_start  {user or toolResult}
        message_end    {...}
        message_start  {assistant}
        message_update {assistant_message_event: a provider StreamEvent}
        message_end    {assistant}
        tool_execution_start / tool_execution_update / tool_execution_end  (if toolUse)
        message_start  {toolResult} / message_end
        turn_end
    agent_end

Events are immutable dataclasses so they can be fanned out to many subscribers safely.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Literal

from minima_harness.ai.events import Event as StreamEvent

if TYPE_CHECKING:
    pass


@dataclass(frozen=True, slots=True)
class AgentStartEvent:
    type: Literal["agent_start"] = "agent_start"


@dataclass(frozen=True, slots=True)
class AgentEndEvent:
    type: Literal["agent_end"] = "agent_end"
    messages: list = field(default_factory=list)  # list[Message]


@dataclass(frozen=True, slots=True)
class TurnStartEvent:
    type: Literal["turn_start"] = "turn_start"


@dataclass(frozen=True, slots=True)
class TurnEndEvent:
    type: Literal["turn_end"] = "turn_end"
    message: Any = None  # AssistantMessage | None
    tool_results: list = field(default_factory=list)  # list[ToolResult]


@dataclass(frozen=True, slots=True)
class MessageStartEvent:
    type: Literal["message_start"] = "message_start"
    message: Any = None  # Message | None


@dataclass(frozen=True, slots=True)
class MessageUpdateEvent:
    """Assistant-only. Wraps a provider streaming event (text/thinking/toolcall delta)."""

    type: Literal["message_update"] = "message_update"
    assistant_message_event: StreamEvent | None = None


@dataclass(frozen=True, slots=True)
class MessageEndEvent:
    type: Literal["message_end"] = "message_end"
    message: Any = None  # Message | None


@dataclass(frozen=True, slots=True)
class ToolExecutionStartEvent:
    type: Literal["tool_execution_start"] = "tool_execution_start"
    tool_call_id: str = ""
    tool_name: str = ""
    args: Any = None  # validated params (pydantic model) | None when blocked/invalid


@dataclass(frozen=True, slots=True)
class ToolExecutionUpdateEvent:
    type: Literal["tool_execution_update"] = "tool_execution_update"
    tool_call_id: str = ""
    partial: Any = None


@dataclass(frozen=True, slots=True)
class ToolExecutionEndEvent:
    type: Literal["tool_execution_end"] = "tool_execution_end"
    tool_call_id: str = ""
    result: Any = None  # ToolResult | None
    is_error: bool = False


AgentEvent = (
    AgentStartEvent
    | AgentEndEvent
    | TurnStartEvent
    | TurnEndEvent
    | MessageStartEvent
    | MessageUpdateEvent
    | MessageEndEvent
    | ToolExecutionStartEvent
    | ToolExecutionUpdateEvent
    | ToolExecutionEndEvent
)


__all__ = [
    "AgentEndEvent",
    "AgentEvent",
    "AgentStartEvent",
    "MessageEndEvent",
    "MessageStartEvent",
    "MessageUpdateEvent",
    "ToolExecutionEndEvent",
    "ToolExecutionStartEvent",
    "ToolExecutionUpdateEvent",
    "TurnEndEvent",
    "TurnStartEvent",
]

File: packages/tui/src/agent/events.ts
/**
 * Agent runtime events — a port of PI's pi-agent-core event taxonomy.
 *
 * Emitted by agentLoop() in a strict order per turn:
 *
 *   agent_start
 *     (per turn)
 *       turn_start
 *       message_start  {user or toolResult}
 *       message_end    {...}
 *       message_start  {assistant}
 *       message_update {assistantMessageEvent: a provider StreamEvent}
 *       message_end    {assistant}
 *       tool_execution_start / tool_execution_update / tool_execution_end  (if toolUse)
 *       message_start  {toolResult} / message_end
 *       turn_end
 *   agent_end
 */

import type { StreamEvent } from "../ai/events.ts";
import type { AssistantMessage, Message } from "../ai/types.ts";
import type { ToolResult } from "./tools.ts";

export interface AgentStartEvent {
  type: "agent_start";
}
export interface AgentEndEvent {
  type: "agent_end";
  messages: Message[];
}
export interface TurnStartEvent {
  type: "turn_start";
}
export interface TurnEndEvent {
  type: "turn_end";
  message: AssistantMessage | null;
  toolResults: ToolResult[];
}
export interface MessageStartEvent {
  type: "message_start";
  message: Message | null;
}
export interface MessageUpdateEvent {
  /** Assistant-only. Wraps a provider streaming event (text/thinking/toolcall delta). */
  type: "message_update";
  assistantMessageEvent: StreamEvent | null;
}
export interface MessageEndEvent {
  type: "message_end";
  message: Message | null;
}
export interface ToolExecutionStartEvent {
  type: "tool_execution_start";
  toolCallId: string;
  toolName: string;
  args: unknown | null; // validated params | null when blocked/invalid
}
export interface ToolExecutionUpdateEvent {
  type: "tool_execution_update";
  toolCallId: string;
  partial: unknown;
}
export interface ToolExecutionEndEvent {
  type: "tool_execution_end";
  toolCallId: string;
  result: ToolResult | null;
  isError: boolean;
}

export type AgentEvent =
  | AgentStartEvent
  | AgentEndEvent
  | TurnStartEvent
  | TurnEndEvent
  | MessageStartEvent
  | MessageUpdateEvent
  | MessageEndEvent
  | ToolExecutionStartEvent
  | ToolExecutionUpdateEvent
  | ToolExecutionEndEvent;

// Factories.
export const agentStart = (): AgentStartEvent => ({ type: "agent_start" });
export const agentEnd = (messages: Message[]): AgentEndEvent => ({ type: "agent_end", messages });
export const turnStart = (): TurnStartEvent => ({ type: "turn_start" });
export const turnEnd = (
  message: AssistantMessage | null,
  toolResults: ToolResult[],
): TurnEndEvent => ({ type: "turn_end", message, toolResults });
export const messageStart = (message: Message | null): MessageStartEvent => ({
  type: "message_start",
  message,
});
export const messageUpdate = (ev: StreamEvent | null): MessageUpdateEvent => ({
  type: "message_update",
  assistantMessageEvent: ev,
});
export const messageEnd = (message: Message | null): MessageEndEvent => ({
  type: "message_end",
  message,
});
export const toolExecutionStart = (
  toolCallId: string,
  toolName: string,
  args: unknown | null,
): ToolExecutionStartEvent => ({ type: "tool_execution_start", toolCallId, toolName, args });
export const toolExecutionUpdate = (
  toolCallId: string,
  partial: unknown,
): ToolExecutionUpdateEvent => ({
  type: "tool_execution_update",
  toolCallId,
  partial,
});
export const toolExecutionEnd = (
  toolCallId: string,
  result: ToolResult | null,
  isError: boolean,
): ToolExecutionEndEvent => ({ type: "tool_execution_end", toolCallId, result, isError });
